### PR TITLE
Add a benchmark.net project

### PR DESCRIPTION
**Pull request by AnthonySteele**
_Summarise the changes this Pull Request makes._

Add a benchmark.net project.
Should i also delete the "PerfTestHarness" project?

_Please include a reference to a GitHub issue if appropriate._

#57

### CODE REVIEW

File: src/Benchmark/BenchmarkStatSending.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            _udpSender.Increment("startup.udp");
+
+            _ipSender = new StatsDPublisher(config, ipTransport);
+            _udpSender.Increment("startup.ip");
+        }
+
+        [Benchmark]
+        public void RunUdp()
+        {
+            _udpSender.Increment("increment.udp");
```
Comment (martincostello) [inline src/Benchmark/BenchmarkStatSending.cs]: It probably makes no _actual_ difference, but maybe they should use strings of equal length so the "overhead" of turning the strings into byte arrays is equal.

File: src/Benchmark/BenchmarkStatSending.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            _udpSender.Timing(TimeSpan.FromMinutes(1), "timer.udp");
```
Comment (martincostello) [inline src/Benchmark/BenchmarkStatSending.cs]: Put this into a static field so the benchmark doesn't include the (tiny) overhead of creating the `TimeSpan`?

File: src/Benchmark/Program.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+using BenchmarkDotNet.Running;
+
+namespace Benchmark
+{
+
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var summary = BenchmarkRunner.Run<BenchmarkStatSending>();
```
Comment (martincostello) [inline src/Benchmark/Program.cs]: Can remove the assignment as it's not used.

File: src/Benchmark/Benchmark.csproj [hunk 4]
```diff
@@ -0,0 +1,21 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <WarningsAsErrors />
```
Comment (martincostello) [inline src/Benchmark/Benchmark.csproj]: Remove if unused?

**Comment by AnthonySteele** [comment]
After these changes, a run on my machine gives

 Method |     Mean |    Error |   StdDev |   Gen 0 | Allocated |
------- |---------:|---------:|---------:|--------:|----------:|
 RunUdp | 299.8 us | 5.750 us | 5.379 us | 44.4336 | 132.34 KB |
  RunIp | 222.2 us | 2.972 us | 2.634 us |  0.9766 |   3.62 KB |

You can compare the effects of test setup to the "master" part of this: https://github.com/justeat/JustEat.StatsD/pull/55#issuecomment-383396253

**Comment by AnthonySteele** [comment]
Should I also delete the "PerfTestHarness" project?

**Comment by martincostello** [comment]
I guess, if this is its replacement.

**Comment by AnthonySteele** [comment]
FYI for docs, my benchmark run procedure is:
````
cd C:\Code\JustEat.StatsD\src\Benchmark
dotnet run -c release
````

**Pull request merged**